### PR TITLE
[TempRValueOpt] Fold added destroy_addrs into loads/copy_addrs.

### PR DESCRIPTION
**Pull request by nate-chandler**
To avoid introducing new copies--which is illegal for move-only values-- don't rewrite `load [take]`s and `copy_addr [take]`s as `load [copy]`s and `copy_addr`s respectively and introduce new `destroy_addr`s after them.  Instead, get the effect of folding such a newly created `destroy_addr` into the preceding rewritten `load [copy]` or `copy_addr`.  Get that effect by neither modifying the `copy_addr [take]` or `load [take]` nor adding a subsequent `destroy_addr`.

An example for each kind (`load [take]` and `copy_addr [take]`):

```
// Input 1 (`load [take]`)
  copy_addr [take] %src to [init] %tmp
  %val = load [take] %src
// Old Output 1
  %val = load [copy] %src
  destroy_addr %src
// New Output 2
  %val = load [take] %src
```

```
// Input 2 (`copy_addr [take]`)
  copy_addr [take] %src to [init] %tmp
  copy_addr [take] %src to [init] %dst
// Old Output 2
  copy_addr %src to [init] %dst
  destroy_addr %src
// New Output 2
  copy_addr [take] %src to [init] %dst
```

rdar://107839979

### CODE REVIEW

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please benchmark

**Comment by nate-chandler** [comment]
@swift-ci please test source compatibility

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please benchmark

**Comment by nate-chandler** [comment]
@swift-ci please test source compatibility

File: lib/SILOptimizer/Transforms/TempRValueElimination.cpp [hunk 1]
```diff
@@ -601,6 +602,28 @@ void TempRValueOptPass::tryOptimizeCopyIntoTemp(CopyAddrInst *copyInst) {
...
+      return true;
+    if (auto *cai = dyn_cast<CopyAddrInst>(lastLoadInst)) {
+      return cai->getSrc() != tempObj || !cai->isTakeOfSrc();
+    }
+    if (auto *li = dyn_cast<LoadInst>(lastLoadInst)) {
+      return li->getOperand() != tempObj ||
+             li->getOwnershipQualifier() != LoadOwnershipQualifier::Take;
+    }
+    return true;
+  }();
```
Comment (eeckstein) [inline lib/SILOptimizer/Transforms/TempRValueElimination.cpp:626]: Maybe you can add an assert here to check if it's a move only type then `needToInsertDestroy` must be false

**Review by eeckstein — APPROVED** [review]
lgtm

**Comment by nate-chandler** [comment]
@swift-ci please test and merge

**Comment by nate-chandler** [comment]
@swift-ci please test linux platform

**Comment by nate-chandler** [comment]
@swift-ci please test macOS platform

**Comment by nate-chandler** [comment]
@swift-ci please test linux platform

File: lib/SILOptimizer/Transforms/TempRValueElimination.cpp [hunk 2]
```diff
@@ -601,6 +602,34 @@ void TempRValueOptPass::tryOptimizeCopyIntoTemp(CopyAddrInst *copyInst) {
...
+  //
+  // This is necessary for correctness: otherwise, copies of move-only values
+  // would be introduced.
+  bool needToInsertDestroy = [&]() {
+    if (!needFinalDeinit)
+      return false;
+    if (lastLoadInst == copyInst)
+      return true;
+    if (auto *cai = dyn_cast<CopyAddrInst>(lastLoadInst)) {
+      auto retval = cai->getSrc() != tempObj || !cai->isTakeOfSrc();
```
Comment (eeckstein) [inline lib/SILOptimizer/Transforms/TempRValueElimination.cpp:619]: nitpick: all these conditions are getting hard to understand. IMO, for clarity it's better to write
```
  if (cai->getSrc() == tempObj && cai->isTakeOfSrc()) {
    // Can reuse the last copy_addr as deinitialization.
    return false;
  } else {
    assert(!tempObj->getType().isMoveOnly() && "introducing copy of move-only value!?");
    return true;
  }
  ```

Comment (nate-chandler) [reply lib/SILOptimizer/Transforms/TempRValueElimination.cpp:619]: https://github.com/apple/swift/pull/65099

**Pull request merged**